### PR TITLE
nixos/nvidia: clean up and add dynamic boost

### PR DESCRIPTION
**Pull request by Gerg-L**
###### Description of changes
Closes: #211300
The NVIDIA module was in need of some TLC so I cleaned it up and re-formatted.

the only actual change in the first commit should be the removal of the `isDeprecated = str: (hasPrefix "nvidia" str) && (str != "nvidia");` check from: #99323 
(2 years old)
and the nvidia prime rename from: #66601
(4 years old)

The second commit is the addition of [NVIDIA Dynamic Boost](https://download.nvidia.com/XFree86/Linux-x86_64/525.53/README/dynamicboost.html) from #211300 

Some more testing would be appreciated

Questions: should I be using `lib.mdDoc` and `lib.options.mkPackageOptionMD` here or not?

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by jian-lin** [comment]
> Questions: should I be using lib.mdDoc and lib.options.mkPackageOptionMD here or not?

I think you should use the `MD` variants because it makes backporting changes to the stable much easier and has no harm.

**Comment by Gerg-L** [comment]
> I think you should use the `MD` variants because it makes backporting changes to the stable much easier and has no harm.

Alright, done.

File: nixos/modules/hardware/video/nvidia.nix [hunk 1]
```diff
@@ -238,6 +244,11 @@ in {
           assertion = cfg.open -> (cfg.package ? open && cfg.package ? firmware);
           message = "This version of NVIDIA driver does not provide a corresponding opensource kernel driver";
         }
+
+        {
```
Comment (Kiskae) [inline nixos/modules/hardware/video/nvidia.nix:248]: Had to look through some of my own WIP code, but powerd is available starting from version `510.39.01`

**Review by Kiskae — CHANGES_REQUESTED** [review]
Other than the required version it looks good to me, although going through this has made me realize the nvidia package itself probably needs a rewrite at some point.

**Comment by Gerg-L** [comment]
> Other than the required version it looks good to me, although going through this has made me realize the nvidia package itself probably needs a rewrite at some point.

Most definitely, I was thinking of rewriting all of `services.xserver` and all the Wayland compositor modules. 

I'm trying to get to it before 22.11 is released

File: nixos/modules/hardware/video/nvidia.nix (hunk 1 above)
Comment (Gerg-L) [reply nixos/modules/hardware/video/nvidia.nix:248]: If it is, it's not documented
[510.39.01](https://download.nvidia.com/XFree86/Linux-x86_64/510.39.01/README/dynamicboost.html)
[510.47.04](https://download.nvidia.com/XFree86/Linux-x86_64/510.47.03/README/dynamicboost.html)

Comment (Kiskae) [reply nixos/modules/hardware/video/nvidia.nix:248]: From `NVIDIA_Changelog`:

     2022-01-11 version 510.39.01

    * Added a new daemon, nvidia-powerd, to provide support for
      the Dynamic Boost feature on supported systems. Dynamic Boost
      balances power between the CPU and the GPU for improved performance.
      For more details, see the "Dynamic Boost on Linux" chapter in the README.

Looks like they added the files, but didn't actually include the documentation for that release. I guess its fine to leave on the original version then.

**Review by Kiskae — APPROVED** [review]

**Review by Kiskae — APPROVED** [review]

**Comment by fpletz** [comment]
> I'm trying to get to it before 22.11 is released

Please send me a copy of the time machine you're using. As mine has been in the shop for aeons I'm not able to fix the bad code I wrote for 17.09. 😆

**Comment by Gerg-L** [comment]
> > I'm trying to get to it before 22.11 is released
> 
> Please send me a copy of the time machine you're using. As mine has been in the shop for aeons I'm not able to fix the bad code I wrote for 17.09. 😆

Fat fingers lol, meant 23.11

**Pull request merged**